cmd/jujud/agent/caasunitinit.go

```diff
@@ -154,22 +165,40 @@ func (c *CAASUnitInitCommand) Run(ctx *cmd.Context) (errOut error) {
 		return errors.Annotatef(err, "failed to make unit tools dir %s",
 			unitPaths.ToolsDir)
 	}
-	if err = c.removeAllFunc(unitPaths.State.BaseDir); err != nil && !os.IsNotExist(err) {
-		return errors.Annotatef(err, "failed to remove unit base dir %s",
-			unitPaths.State.BaseDir)
+	if c.Args.Upgrade {
+		if c.Args.CharmDir != "" {
+			if err = c.removeAllFunc(unitPaths.State.CharmDir); err != nil && !os.IsNotExist(err) {
+				return errors.Annotatef(err, "failed to remove unit charm dir %s",
+					unitPaths.State.CharmDir)
+			}
+		}
+	} else {
+		if err = c.removeAllFunc(unitPaths.State.BaseDir); err != nil && !os.IsNotExist(err) {
+			return errors.Annotatef(err, "failed to remove unit base dir %s",
+				unitPaths.State.BaseDir)
+		}
+		if err = c.mkdirAllFunc(unitPaths.State.BaseDir, 0775); err != nil {
+			return errors.Annotatef(err, "failed to make unit base dir %s",
+				unitPaths.State.BaseDir)
+		}
```

Comment:
IIAMN, the current deployer unpacks the charm into an existing directory using a manifest to figure out what files need adding/removing/updating etc. Can we check that we can't continue to use that approach?

Reply:
This is the caas unit init (so it didn't use the deployer before, and currently doesn't use it), for now I'd rather continue doing what we already did, which is compress the deployed charm, and unpack it on the other end. We can revisit this at some point.